### PR TITLE
Reduce span highlighted code in unused_variables lint 

### LINKED ISSUE #50472: Reduce span highlighted code in unused_variables lint
Let's say I wrote this critical piece of code:

```rust
fn main() {
    let mut i = 42;
}
```

rustc gives me the following output:

```
warning: unused variable: `i`
 --> src/main.rs:2:9
  |
2 |     let mut i = 42;
  |         ^^^^^ help: consider using `_i` instead
  |
  = note: #[warn(unused_variables)] on by default

warning: variable does not need to be mutable
 --> src/main.rs:2:9
  |
2 |     let mut i = 42;
  |         ----^
  |         |
  |         help: remove this `mut`
  |
  = note: #[warn(unused_mut)] on by default
```

Which is correct -- except for one little detail. Which is crucial, if you are not just any developer but actually _rustfix_.

You see, the `mut ` in `mut i` is highlighted, but the replacement is just for the binding name, not the mutability. (Why not the mutability? Because the mutability is covered by the next lint we trigger, `unused_mut`.) And _rustfix_ will rightfully complain if we try to replace parts of the code that we have already replaced.

tl;dr We need to change the span in `unused_variables` to only cover the `i`

### PR DESCRIPTION
**Pull request by csmoe**
Fixes #50472 
- [X] reduce var span
- [ ] mark applicable
 
Before:
```
mut unused_mut_var
^^^^^^^^^^^^^^^^^^
```
After:
```
mut unused_mut_var
    ^^^^^^^^^^^^^^
```

### CODE REVIEW

File: src/librustc/middle/liveness.rs [hunk 1]
```diff
@@ -1539,8 +1545,11 @@ impl<'a, 'tcx> Liveness<'a, 'tcx> {
                     self.assigned_on_exit(ln, var).is_some()
                 };
 
-                let suggest_underscore_msg = format!("consider using `_{}` instead",
-                                                     name);
+                let (suggest_underscore_msg, app) = (
```
Comment (oli-obk) [inline src/librustc/middle/liveness.rs]: Please change this to two separate let bindings

File: src/test/ui/lint/issue-47390-unused-variable-in-struct-pattern.rs [hunk 2]
```diff
@@ -35,6 +35,10 @@ fn main() {
         endless_and_singing: true
     };
 
+    let mut mut_unused_var = 1;
```
Comment (oli-obk) [inline src/test/ui/lint/issue-47390-unused-variable-in-struct-pattern.rs]: Not sure if that works, but you can try adding a `// run-rustfix` comment to the top of the file. That'll show us the file after applying the suggestions

File: src/librustc/middle/liveness.rs [hunk 3]
```diff
@@ -1398,7 +1399,11 @@ fn check_local<'a, 'tcx>(this: &mut Liveness<'a, 'tcx>, local: &'tcx hir::Local)
         },
         None => {
             this.pat_bindings(&local.pat, |this, ln, var, sp, id| {
-                this.warn_about_unused(sp, id, ln, var);
```
Comment (oli-obk) [inline src/librustc/middle/liveness.rs]: Instead of the if let you could prepend this line with `let sp = local.pat.simple_span.unwrap_or(sp);`

**Review by oli-obk — COMMENTED** [review]
Just some nits

**Comment by oli-obk** [comment]
@killercup what actions are needed here?

**Comment by csmoe** [comment]
@oli-obk I had talked with @killercup on discord, the failed test will be fixed once a new version rustfix published.

**Comment by killercup** [comment]
In addition to the `// run-rustfix` line, you'll also need to a the a corresponding `.fixed` file,  that contains the fixed code. This is probablematic right now, because the rustfix compiletest uses needs to be updated so these fixes work (I'll publish a new version and open a PR here soon).

We should land this right now, but without the rustfix test mode and the applicability flag. After a rustfix update, we can then add the applicability flag and "prove" it's working by adding the rustfix tests.

**Comment by oli-obk** [comment]
makes sense.

@bors r+

**Pull request merged**